### PR TITLE
Type inference: Propagate declare type information

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**

### CODE REVIEW

File: python-frontend/src/test/java/org/sonar/python/tree/CallExpressionImplTest.java [hunk 1]
```diff
@@ -102,4 +111,95 @@ public void not_callable_callee_symbol_type() {
...
+      "  x"
+    ).type()).isEqualTo(anyType());
+
+    assertThat(lastExpression(
+      "def foo(param):",
+      "  x = param.capitalize()",
+      "  x"
+    ).type()).isEqualTo(anyType());
+
+    // TODO: handle member access dependencies for declared type
```
Comment (guillaume-dequenne-sonarsource) [inline python-frontend/src/test/java/org/sonar/python/tree/CallExpressionImplTest.java:135]: What would be required for this? I assume we'd need to link the last `capitalize` called to the declared type of `param.capitalize`? Do we have a ticket for that?

File: python-frontend/src/test/java/org/sonar/python/tree/CallExpressionImplTest.java [hunk 2]
```diff
@@ -102,4 +111,95 @@ public void not_callable_callee_symbol_type() {
...
+    assertThat(lastExpression(
+      "from typing import Union",
+      "class A:",
+      "  def meth() -> Union[int, str]: ...",
+      "def foo(param: A):",
+      "  x = param.meth()",
+      "  x"
+    ).type()).isEqualTo(new DeclaredType(union, typeArgs));
+
+    // TODO: handle UnionType of annotation coming from typeshed
```
Comment (guillaume-dequenne-sonarsource) [inline python-frontend/src/test/java/org/sonar/python/tree/CallExpressionImplTest.java:164]: Should we create & reference a ticket for that here?

File: python-frontend/src/test/java/org/sonar/python/tree/CallExpressionImplTest.java (hunk 1 above)
Comment (andrea-guarino-sonarsource) [reply python-frontend/src/test/java/org/sonar/python/tree/CallExpressionImplTest.java]: done: https://jira.sonarsource.com/browse/SONARPY-781

File: python-frontend/src/test/java/org/sonar/python/tree/CallExpressionImplTest.java (hunk 2 above)
Comment (andrea-guarino-sonarsource) [reply python-frontend/src/test/java/org/sonar/python/tree/CallExpressionImplTest.java]: done: https://jira.sonarsource.com/browse/SONARPY-782

**Pull request merged**